components/script/dom/htmlcanvaselement.rs

```diff
@@ -202,6 +192,48 @@ impl HTMLCanvasElement {
         }
     }
 
+    pub fn get_or_init_webgl2_context(&self,
+                                      cx: *mut JSContext,
+                                      attrs: Option<HandleValue>)
+                                      -> Option<DomRoot<WebGL2RenderingContext>> {
+        if !PREFS.is_webgl2_enabled() {
+            return None
+        }
+        if self.context.borrow().is_none() {
+            let window = window_from_node(self);
+            let size = self.get_size();
+            let attrs = Self::get_gl_attributes(cx, attrs)?;
+            let maybe_ctx = WebGL2RenderingContext::new(&window, self, size, attrs);
+
+            *self.context.borrow_mut() = maybe_ctx.map( |ctx| CanvasContext::WebGL2(Dom::from_ref(&*ctx)));
+        }
+
+        if let Some(CanvasContext::WebGL2(ref context)) = *self.context.borrow() {
+            Some(DomRoot::from_ref(&*context))
+        } else {
+            None
+        }
+    }
+
+    #[allow(unsafe_code)]
+    fn get_gl_attributes(cx: *mut JSContext, attrs: Option<HandleValue>) -> Option<GLContextAttributes> {
+        if let Some(webgl_attributes) = attrs {
```

Comment:
nit: This can read nicer like:

```rust

let webgl_attributes = match attrs {
    Some(attrs) => attrs,
    None => return Some(GlContextAttributes::default()),
};

// ...
```